cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -0,0 +1,141 @@
+
+  /**
+   * Gets the uninitialized local variable corresponding to this node, if
+   * any.
+   */
+  LocalVariable asUninitialized() {
+    result = this.(UninitializedInstruction).getLocalVariable()
+  }
+
+  /**
+   * Gets an upper bound on the type of this node.
+   */
+  Type getTypeBound() { result = getType() }
+}
+
+/**
+ * An expression, viewed as a node in a data flow graph.
+ */
+class ExprNode extends Node {
+  ExprNode() { getAST() instanceof Expr }
+  Expr getExpr() { result = getAST() }
+}
+
+/**
+ * The value of a parameter at function entry, viewed as a node in a data
+ * flow graph.
+ */
+class ParameterNode extends Node, InitializeParameterInstruction {
+  /**
+   * Holds if this node is the parameter of `c` at the specified (zero-based)
+   * position. The implicit `this` parameter is considered to have index `-1`.
+   */
+  predicate isParameterOf(Function f, int i) {
+    f.getParameter(i) = getParameter()
+  }
+}
+
+/**
+ * The value of an uninitialized local variable, viewed as a node in a data
+ * flow graph.
+ */
+class UninitializedNode extends Node, UninitializedInstruction {
+}
+
+/**
+ * A node associated with an object after an operation that might have
+ * changed its state.
+ *
+ * This can be either the argument to a callable after the callable returns
+ * (which might have mutated the argument), or the qualifier of a field after
+ * an update to the field.
+ *
+ * Nodes corresponding to AST elements, for example `ExprNode`, usually refer
+ * to the value before the update with the exception of `ClassInstanceExpr`,
+ * which represents the value after the constructor has run.
+ */
+abstract class PostUpdateNode extends Node {
+  /**
+   * Gets the node before the state update.
+   */
+  abstract Node getPreUpdateNode();
+}
+
+class StoreDestinationAsPostUpdateNode extends PostUpdateNode {
+  StoreInstruction si;
+  StoreDestinationAsPostUpdateNode() {
+    this = si.getDestinationAddress()
+  }
+  
+  override Node getPreUpdateNode() {
+    result = si.getDestinationAddress()
+  }
+}
+
+/**
+ * Gets the `Node` corresponding to `e`.
+ */
+ExprNode exprNode(Expr e) { result.getExpr() = e }
+
+/**
+ * Gets the `Node` corresponding to the value of `p` at function entry.
+ */
+ParameterNode parameterNode(Parameter p) { result.getParameter() = p }
+
+/**
+ * Gets the `Node` corresponding to the value of an uninitialized local
+ * variable `v`.
+ */
+UninitializedNode uninitializedNode(LocalVariable v) {
+  result.getLocalVariable() = v
+}
+
+/**
+ * Holds if data flows from `nodeFrom` to `nodeTo` in exactly one local
+ * (intra-procedural) step.
+ */
+predicate localFlowStep(Node nodeFrom, Node nodeTo) {
+  nodeTo.(CopyInstruction).getSourceValue() = nodeFrom or
+  nodeTo.(PhiInstruction).getAnOperand().getDefinitionInstruction() = nodeFrom or
+  // Treat all conversions as flow, even conversions between different numeric types.
+  nodeTo.(ConvertInstruction).getOperand() = nodeFrom
```

Comment:
This is beautiful, and it shows the power of the IR: a step in the IR is essentially a data-flow step.